### ISSUE TITLE
Explicit prioritization for EntityRuler matches

### ISSUE THREAD

**Issue by antelope-admin** [explosion/spaCy#7028]
Current documentation [mentions](https://spacy.io/usage/rule-based-matching#entityruler-usage) the behaviour of EntityRuler in case of overlapping matches: 

>  If any matches were to overlap, the pattern matching most tokens takes priority. If they also happen to be equally long, then the match occurring first in the Doc is chosen.

However, there is no mention anywhere of cases where the matches occur at same place in document, _and_ are of same length.

Here's a realistic example of when this would happen. We have a few acronyms that are caught exactly as they appear (in this example I included "ITM" which means 'in the money'), and we want anything that is not explicit but has the morphological shape of `XXX` or `XXXX` to be marked as a ticker symbol (e.g. `AAPL`). 

It looks like currently how things get tagged with the following definitions is not user-specifiable nor deterministic:
 

```python
nlp = spacy.load("en_core_web_trf")

## Define custom rule-based matching logic
ruler = nlp.add_pipe("entity_ruler", before="ner")

patterns = [
    dict(
        label="OPTIONS_SLANG",
        pattern="ITM",
    ),
    dict(
        label="TICKER",
        pattern=[
            {"SHAPE": "XXXX"}
        ],
    ),
    dict(
        label="TICKER",
        pattern=[
            {"SHAPE": "XXX"}
        ],
    ),
]

ruler.add_patterns(patterns)

```

this sometimes results in the "right" parse (e.g. `ITM` would be tagged as `OPTIONS_SLANG`) but more often than not ends up applying the more generic `shape: XXX` pattern and tags everything as `TICKER`: 

```python
example = "buying way too many ITM TSLA options. we're not in IPO times here. ROI is low"
doc = nlp(example)
print([
    (ent.text, ent.label_)
    for ent in doc.ents
])
[('ITM', 'TICKER'), ('TSLA', 'TICKER'), ('IPO', 'TICKER'), ('ROI', 'TICKER')]
```

Further, it seems that there are some non-local dependencies at work here; e.g. if I change the above text slightly to remove last match (`ROI`), things seem to work more "correctly" consistently:

```python
example = "buying way too many ITM TSLA options. we're not in IPO times here."
doc = nlp(example)
[
    (ent.text, ent.label_)
    for ent in doc.ents
]
[('ITM', 'OPTIONS_SLANG'), ('TSLA', 'TICKER'), ('IPO', 'TICKER')]
```

TLDR:

1. It would be nice to be able to specify a priority / weight, or otherwise at least have some default behavior where first pattern / dict rule defined takes precedence.
2. I'm not sure the current logic makes sense regardless; adding/removing additional text when using entityruler before NER in the pipeline should not impact the entityruler-based labels I would think?

## Your Environment

- Operating System: OSX
- Python Version Used: 3.8.1
- spaCy Version Used: 3.x.0 (latest)
- Environment Information:

**Comment by adrianeboyd** [comment]
I think the order of the returned matches depends on some fairly complicated internal `Matcher` state details (whether there are open partial matches, etc.), which is why changing the text can change the order of the returned matches.

I think it would be easy to do something like sort the matches alphabetically by label in the `EntityRuler` so that it's at least more predictable from the user's perspective. Making it more configurable gets tricky because of serialization, but we could consider supporting a custom registered sorting function perhaps?

As an alternative, you can extend `EntityRuler` with a custom component name and override `set_annotations` to get more customized behavior.

**Comment by craigmarvelley** [comment]
I have a similar issue with a slightly different use case, I hope you don't mind me adding it in in case it's useful:

I'm also matching tickers. These are finite, and in many cases our system is aware of them — in these cases, we'd want to associate them with our ID, e.g.:

{"label":"TICKER","pattern":"HMRK","id":"1234"}

And then to cover any as-yet-unknown tickers, we'd define a regex, e.g.:

{"label":"TICKER","pattern":[{"TEXT":{"REGEX":"^[A-Z\\d]+$"}}]}

I'd prefer that the more explicit pattern matched first, so that we could retrieve the ID in that case. However it seems that even if there is an exact match, the EntityRuler matches on the regex pattern instead, so we do not get the ID. As the OP suggested I tried playing with the order of the patterns etc. but it doesn't seem to make a difference, which now makes sense from your comment @adrianeboyd.

We'll probably get around this by having two separate models for explicit patterns and generic ones, but I thought I'd mention it.

**Comment by adrianeboyd** [comment]
So this isn't going to be addressed in the current entity ruler directly, we've add a new `span_ruler` component that is more configurable to cover this case. This is planned for v3.3.1 and you can define custom sorting and filtering before the matches are assigned to `doc.ents` and/or `doc.spans`.
